pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,816 @@
+	}
+	c.trafficControlInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addTC,
+			UpdateFunc: c.updateTC,
+			DeleteFunc: c.deleteTC,
+		},
+		resyncPeriod,
+	)
+	c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addPod,
+			UpdateFunc: c.updatePod,
+			DeleteFunc: c.deletePod,
+		},
+		resyncPeriod,
+	)
+	c.namespaceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addNamespace,
+			UpdateFunc: c.updateNamespace,
+			DeleteFunc: c.deleteNamespace,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+func (c *Controller) matchPod(pod *v1.Pod, to *v1alpha2.AppliedTo) bool {
+	if to.NamespaceSelector != nil {
+		namespace, _ := c.namespaceLister.Get(pod.Namespace)
+		nsSelector, _ := metav1.LabelSelectorAsSelector(to.NamespaceSelector)
+		if nsSelector.Matches(labels.Set(namespace.GetLabels())) {
+			return true
+		}
+	}
+
+	podSelector, _ := metav1.LabelSelectorAsSelector(to.PodSelector)
+	if podSelector.Matches(labels.Set(pod.Labels)) {
+		return true
+	}
+	return false
+}
+
+func (c *Controller) matchNamespace(namespace *v1.Namespace, to *v1alpha2.AppliedTo) bool {
+	if to.NamespaceSelector != nil {
+		nsSelector, _ := metav1.LabelSelectorAsSelector(to.NamespaceSelector)
+		if nsSelector.Matches(labels.Set(namespace.GetLabels())) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Controller) filterAffectedTCsByPod(pod *v1.Pod) sets.String {
+	affectedTCs := sets.NewString()
+	allTCs, _ := c.trafficControlLister.List(labels.Everything())
+	for _, tc := range allTCs {
+		if c.matchPod(pod, &tc.Spec.AppliedTo) {
+			affectedTCs.Insert(tc.GetName())
+		}
+	}
+	return affectedTCs
+}
+
+func (c *Controller) filterAffectedTCsByNS(namespace *v1.Namespace) sets.String {
+	affectedTCs := sets.NewString()
+	allTCs, _ := c.trafficControlLister.List(labels.Everything())
+	for _, tc := range allTCs {
+		if c.matchNamespace(namespace, &tc.Spec.AppliedTo) {
+			affectedTCs.Insert(tc.GetName())
+		}
+	}
+	return affectedTCs
+}
+
+func (c *Controller) addPod(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	if pod.Spec.HostNetwork {
+		return
+	}
+	affectedTCs := c.filterAffectedTCsByPod(pod)
+	if len(affectedTCs) != 0 {
+		klog.V(2).InfoS("Processing TrafficControls affected by Pod ADD event", "Pod", klog.KObj(pod))
+		for tc := range affectedTCs {
+			c.queue.Add(tc)
+		}
+	}
+}
+
+func (c *Controller) updatePod(oldObj interface{}, obj interface{}) {
+	oldPod := oldObj.(*v1.Pod)
+	pod := obj.(*v1.Pod)
+	if oldPod.Spec.HostNetwork && pod.Spec.HostNetwork {
+		return
+	}
+	if !reflect.DeepEqual(pod.GetLabels(), oldPod.GetLabels()) {
+		oldPodAffectedTCs := c.filterAffectedTCsByPod(oldPod)
+		newPodAffectedTCs := c.filterAffectedTCsByPod(pod)
+		if len(oldPodAffectedTCs) == 0 && len(newPodAffectedTCs) == 0 {
```

Comment:
this it not really needed, `SymmetricDifferenceString` will get empty result in such case.